### PR TITLE
Add e2e test which verifies traffic policies and firewall in services

### PR DESCRIPTION
**Pull request by manuelbuil**
#### Proposed Changes ####

New e2e test to verify services externaltrafficpolicy, internaltrafficpolicy and loadBalancerSourceRanges

#### Types of Changes ####

New e2e test

#### Verification ####

Run go test

#### User-Facing Change ####

```release-note

```

### CODE REVIEW

**Review by thomasferrandiz — APPROVED** [review]

**Review by VestigeJ — APPROVED** [review]

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go [hunk 1]
```diff
@@ -0,0 +1,363 @@
+// This test verifies:
```
Comment (dereknola) [inline tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:1]: Note: I hand wrote these corrections, so spacing might be bad, don't just commit them via GH :smile:

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+			return len(externalIPs), nil
+		}, "25s", "5s").Should(Equal(2), "external IP count not equal to 2")
+
+		Eventually(func() (bool, error) {
+			externalIPs, _ := e2e.FetchExternalIPs(kubeConfigFile, lbSvcExt)
+			if len(externalIPs) != 1 {
+				return false, nil
+			}
+			return externalIPs[0] == serverNodeIP, nil
+		}, "25s", "5s").Should(BeTrue(), "external IP count not equal to 1 or external IP does not match serverNodeIP")
```
Comment (dereknola) [inline tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:136]:
```suggestion
		Eventually(func(g Gomega) (bool, error) {
			externalIPs, _ := e2e.FetchExternalIPs(kubeConfigFile, lbSvcExt)
			g.Expect(externalIPs).To(HaveLen(1), "more than 1 exernalIP found")
			g.Expect(externalIPs[0]).To(Equal(serverNodeIP),"external IP does not match servernodeIP")
		}, "25s", "5s").Should(Succeed())
```

Comment (dereknola) [reply tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:136]: We can use the gomega [Assertions In Eventually](https://onsi.github.io/gomega/#category-3-making-assertions-eminem-the-function-passed-into-codeeventuallycode) feature to standardize the checks

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+			}
+			Eventually(func() (bool) {
+				cmd := "curl -s --max-time 5 " + externalIP + ":82/ip"
+				_, err := e2e.RunCommand(cmd)
+				if err != nil && strings.Contains(err.Error(), "exit status") {
+					// Treat exit status as a successful condition
+					return true
+				}
+				return false
+			}, "40s", "5s").Should(BeTrue())
```
Comment (dereknola) [inline tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:176]: Use [gomega error matcher](https://onsi.github.io/gomega/#matcherrorexpected-interface)
```suggestion
			Eventually(func() error {
				cmd := "curl -s --max-time 5 " + externalIP + ":82/ip"
				_, err := e2e.RunCommand(cmd)
				return err
			}, "40s", "5s").Should(MatchError(ContainSubstring("exit status")))
```

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+						clientPod1IP = pod.IP
+					} else {
+						clientPod2 = pod.Name
+						clientPod2Node = pod.Node
+						clientPod2IP = pod.IP
+					}
+				}
+			}
+			return serverNodeName != "" && clientPod1 != "" && clientPod2 != "" && clientPod1Node != "" && clientPod2Node != "", nil
+		}, "25s", "5s").Should(BeTrue(), "All pod names and nodes should be non-empty")
```
Comment (dereknola) [inline tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:230]: You should be able to rewrite this using the `g Gomega` feature like the above comment, don't embed a complex series of checks as a single return statement, have each one be check with some `g.Expect(serverNodeName).NotTo(BeEmpty)` etc.

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+			// Verify connectivity from nodes[1] fails because we did not pass its IP to the loadBalancerSourceRanges
+			Eventually(func() (bool) {
+				cmd := "curl -s --max-time 5 " + node.InternalIP + ":82"
+				_, err := e2e.RunCmdOnNode(cmd, nodes[1].Name)
+				if err != nil && strings.Contains(err.Error(), "exit status") {
+					// Treat exit status as a successful condition
+					return true
+				}
+				return false
+			}, "40s", "5s").Should(BeTrue())
```
Comment (dereknola) [inline tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:345]: Same as above, rewrite using the [gomega error matcher ](https://onsi.github.io/gomega/#matcherrorexpected-interface)checks

**Review by dereknola — CHANGES_REQUESTED** [review]

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+	"github.com/k3s-io/k3s/tests/e2e"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+// Valid nodeOS: bento/ubuntu-24.04, opensuse/Leap-15.6.x86_64
+var nodeOS = flag.String("nodeOS", "bento/ubuntu-24.04", "VM operating system")
+var serverCount = flag.Int("serverCount", 1, "number of server nodes")
+var agentCount = flag.Int("agentCount", 1, "number of agent nodes")
+var hardened = flag.Bool("hardened", false, "true or false")
```
[the remark this answers is not in the document]
Comment (manuelbuil) [reply tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:25]: good

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+			}
+		}
+
+		// Verify there is only one external-ip and it is matching the node IP
+		lbSvc := "nginx-loadbalancer-svc"
+		lbSvcExt := "nginx-loadbalancer-svc-ext"
+		Eventually(func() (int, error) {
+			externalIPs, _ := e2e.FetchExternalIPs(kubeConfigFile, lbSvc)
+			return len(externalIPs), nil
+		}, "25s", "5s").Should(Equal(2), "external IP count not equal to 2")
```
[the remark this answers is not in the document]
Comment (manuelbuil) [reply tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:128]: Thanks, I did not know about the `HaveLen`

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go (hunk 2 above)
Comment (manuelbuil) [reply tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:136]: thanks! Your suggestion looks better :)

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go (hunk 4 above)
Comment (manuelbuil) [reply tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:230]: When passing `func(g Gomega)` and checking variables along the function, there isn't really a need to return anything, right?

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go (hunk 5 above)
Comment (manuelbuil) [reply tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go:345]: done!

**Review by dereknola — APPROVED** [review]

File: tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go (hunk 4 above)
Comment (dereknola) [reply tests/e2e/svcpoliciesandfirewall/svcpoliciesandfirewall_test.go]: Correct, nothing needs to be returned.

**Review by thomasferrandiz — APPROVED** [review]

**Pull request merged**